### ISSUE TITLE
AssertionError occurs in function show_result() of inference.py when class_names has only one class

### ISSUE THREAD

**Issue by xuhuasheng** [open-mmlab/mmdetection#1570]
**Checklist**
1. I have searched related issues but cannot get the expected help.
2. The bug has not been fixed in the latest version.

**Describe the bug**
the bug is in `mmdet/apis/inference.py -line 116` `assert isinstance(class_names, (tuple, list))` 
the bug will arise when there is only one class in `class_names`, which will raise `AssertionError`

**Environment**
 - OS: [Ubuntu 18.04.2]
 - GCC [5.5.0]
 - PyTorch version [1.1.0]
- GPU model [1080Ti]
- CUDA9.0and CUDNN7.4

**Error traceback**
```
raceback (most recent call last):
  File "inference_demo.py", line 81, in <module>
    main()
  File "inference_demo.py", line 75, in main
    show_result(img_fullFileName, result, model.CLASSES, wait_time=-1, show=False, out_file=save_img_fullFileName)
  File "/home/watson/anaconda3/envs/pytorch/mmdetection/mmdet/apis/inference.py", line 116, in show_result
    assert isinstance(class_names, (tuple, list)) 
AssertionError
```

**Bug fix**
the bug above will be fixed by modifing `mmdet/apis/inference.py , line 116`
 modify 
`assert isinstance(class_names, (tuple, list))`
 to
`assert isinstance(class_names, (tuple, list, str))`

**Comment by hellock** [comment]
Thanks! The behavior is intended. If there is only one class, the class names should be `['object']` or `('object', )`.

**Comment by xuhuasheng** [comment]
> Thanks! The behavior is intended. If there is only one class, the class names should be `['object']` or `('object', )`.

Thank your reply! 
In my project, there is only one class, so I have modified `CLASS = ('myObject', )` in `datasets/coco.py`, but it will also raise the same `AssertionError` in `assert isinstance(class_names, (tuple, list)) `.
when I modified `assert isinstance(class_names, (tuple, list, str))`, it worked.

**Comment by yhcao6** [comment]
Hi, I just test `show_result` when class_names has only one class, there is no error occur. Here is my change and test pipeline:
1. `CLASSES = (person, )` in [coco.py](https://github.com/open-mmlab/mmdetection/blob/618dca0895f5f4ede19f5feebb064648e128e12e/mmdet/datasets/coco.py#L11).
2. `num_classes=2` in [cfg](https://github.com/open-mmlab/mmdetection/blob/618dca0895f5f4ede19f5feebb064648e128e12e/configs/faster_rcnn_r50_fpn_1x.py#L40).
3. Train one iteration and save the model.
4. This is the `CLASSES` saved in checkpoint:
```
 ...
 'CLASSES': ('person',),
 'epoch': 1,
 'iter': 0,
 'mmcv_version': '0.2.13',
 'time': 'Wed Oct 23 14:56:18 2019'}
```

5. Test `show_result` function, here is my script:
```
import os.path as osp
import sys

sys.path.insert(0, osp.abspath(osp.join(__file__, '../../')))

from mmdet.apis import init_detector, inference_detector, show_result

config_file = 'configs/test/faster_rcnn_r50_fpn_1x.py'
checkpoint_file = 'work_dirs/faster_rcnn_r50_fpn_1x/epoch_1.pth'

# build the model from a config file and a checkpoint file
model = init_detector(config_file, checkpoint_file, device='cuda:0')

# test a single image and show the results
img = 'demo/demo.jpg'
result = inference_detector(model, img)
# or save the visualization results to image files
show_result(img, result, model.CLASSES, show=False, out_file='img/result.jpg')
```

Could you provide a more concrete example where error will occur?

**Comment by xuhuasheng** [comment]
> Hi, I just test `show_result` when class_names has only one class, there is no error occur. Here is my change and test pipeline:
> 
> 1. `CLASSES = (person, )` in [coco.py](https://github.com/open-mmlab/mmdetection/blob/618dca0895f5f4ede19f5feebb064648e128e12e/mmdet/datasets/coco.py#L11).
> 2. `num_classes=2` in [cfg](https://github.com/open-mmlab/mmdetection/blob/618dca0895f5f4ede19f5feebb064648e128e12e/configs/faster_rcnn_r50_fpn_1x.py#L40).
> 3. Train one iteration and save the model.
> 4. This is the `CLASSES` saved in checkpoint:
> 
> ```
>  ...
>  'CLASSES': ('person',),
>  'epoch': 1,
>  'iter': 0,
>  'mmcv_version': '0.2.13',
>  'time': 'Wed Oct 23 14:56:18 2019'}
> ```
> 
> 1. Test `show_result` function, here is my script:
> 
> ```
> import os.path as osp
> import sys
> 
> sys.path.insert(0, osp.abspath(osp.join(__file__, '../../')))
> 
> from mmdet.apis import init_detector, inference_detector, show_result
> 
> config_file = 'configs/test/faster_rcnn_r50_fpn_1x.py'
> checkpoint_file = 'work_dirs/faster_rcnn_r50_fpn_1x/epoch_1.pth'
> 
> # build the model from a config file and a checkpoint file
> model = init_detector(config_file, checkpoint_file, device='cuda:0')
> 
> # test a single image and show the results
> img = 'demo/demo.jpg'
> result = inference_detector(model, img)
> # or save the visualization results to image files
> show_result(img, result, model.CLASSES, show=False, out_file='img/result.jpg')
> ```
> 
> Could you provide a more concrete example where error will occur?

How to checkout  the `CLASSES` saved in checkpoint?

**Comment by yhcao6** [comment]
```
import torch
cpt = torch.load(cpt_path)
print(cpt['meta']['CLASSES'])
```

**Comment by xuhuasheng** [comment]
Sorry, it's my fault. I checked the `CLASSES` saver in checkpoint file, I found it's a `str` instead of `tuple`.

My mistake:
I modified `CLASSES = (object)` in coco.py before training. but I corrected it to `CLASSES = (object, )`  when I realized it's wrong, before inferencing, whichi will raise AssertError.

My issue has been solved. Thank you very much!

**Comment by hande6688** [comment]
@xuhuasheng How to label a single category with labelme and instance segmentation? There are cat, cat, cat, cat0, cat1, cat2 or cat-0, cat-1, cat-2, etc. which annotation method is used?thank you!

I use cat-0, cat-1, cat-2 for annotation, and use labelme2coco to convert annotations.json, where the annotation is [], what's the matter?
